libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -122,78 +152,109 @@ impl Replacements {
         Some(replacements)
     }
 
-    fn insert(&mut self, k: ExprPat, v: Expr) -> Option<Expr> {
-        self.map.insert(k, v)
+    fn insert(&mut self, k: &Rc<ExprPat>, v: Rc<Expr>) -> Option<Rc<Expr>> {
+        self.map.insert(hash(k.as_ref()), v)
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::{parse_expression, parse_expression_pattern, scan};
+
+    fn parse_rule(prog: &str) -> ExprPat {
+        let (expr, _) = parse_expression_pattern(scan(prog));
+        expr
+    }
+
+    fn parse_expr(prog: &str) -> Expr {
+        match parse_expression(scan(prog)) {
+            (Stmt::Expr(expr), _) => expr,
+            _ => unreachable!(),
+        }
```

Comment:
I don't understand the meaning for this statement. Shouldn't you use if let since all you're doing is returning the enum variant?

Reply:
No, we're returning data inner to the enum variant which is inner to the tuple returned by `parse_Expression`. Even if we use `if let`, we have to handle the default (unreachable case)